### PR TITLE
binlog-filter: add new event type

### LINKED ISSUE #656: Add new event type for binlog-filter.
## Feature Request

**Is your feature request related to a problem? Please describe:**

As title.
**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve?

Issue Number: close #656 
ref https://github.com/pingcap/tiflow/issues/6160 

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

 - Increased code complexity

Related changes

### CODE REVIEW

File: pkg/binlog-filter/filter_test.go [hunk 1]
```diff
@@ -14,10 +14,9 @@
 package filter
 
 import (
-	"testing"
-
 	. "github.com/pingcap/check"
 	selector "github.com/pingcap/tidb-tools/pkg/table-rule-selector"
+	"testing"
```
Comment (okJiang) [inline pkg/binlog-filter/filter_test.go]: wrong place

**Review by okJiang — APPROVED** [review]
LGTM

File: pkg/binlog-filter/filter.go [hunk 2]
```diff
@@ -64,9 +64,41 @@ const (
...
 	AlertTable     EventType = "alter table"
+
+	CreateSchema                  EventType = "create schema" // alias of CreateDatabase
+	DropSchema                    EventType = "drop schema"   // alias of DropDatabase
+	AddColumn                     EventType = "add column"
+	DropColumn                    EventType = "drop column"
+	AddIndex                      EventType = "add index" // alias of CreateIndex
+	AddForeignKey                 EventType = "add foreign key"
+	DropForeignKey                EventType = "drop foreign key"
+	ModifyColumn                  EventType = "modify column"
```
Comment (lance6716) [inline pkg/binlog-filter/filter.go]: It only lists a part of DDL, for example. you missed ALTER TABLE CHANGE COLUMN.

How can we use these new types?

Comment (asddongmen) [reply pkg/binlog-filter/filter.go]: 1. No, I didn't, all event types TiCDC want to support are aboved.
2. TiCDC's will allow user to specify these new added even type to filter rule.

File: pkg/binlog-filter/filter_test.go (hunk 1 above)
Comment (asddongmen) [reply pkg/binlog-filter/filter_test.go]: thanks, good catch.

File: pkg/binlog-filter/filter.go (hunk 2 above)
Comment (lance6716) [reply pkg/binlog-filter/filter.go]: the problem is even you, don't know the the meaning of these types. ALTER TABLE CHANGE COLUMN will generate a "modify column" DDL job, as well as ALTER TABLE MODIFY COLUMN. So we can't count on that user will understand their meaning.

These types derive from TiDB's internal type of ddl jobs, I think configuration should not rely on internal information.

**Review by lance6716 — CHANGES_REQUESTED** [review]
need discuss with PM

File: pkg/binlog-filter/filter.go [hunk 3]
```diff
@@ -63,24 +63,33 @@ const (
...
+	AlterTable     EventType = "alter table"
+
+	CreateSchema EventType = "create schema" // alias of CreateDatabase
+	DropSchema   EventType = "drop schema"   // alias of DropDatabase
+	AddIndex     EventType = "add index"     // alias of CreateIndex
 	// if need, add more	AlertTableOption     = "alert table option"
 
 	NullEvent EventType = ""
+	// InvalidEvent is used to indicate invalid event type
+	InvalidEvent EventType = "invalid event"
```
Comment (lance6716) [inline pkg/binlog-filter/filter.go]: can we delete `NoneEvent` now?

File: pkg/binlog-filter/filter.go [hunk 4]
```diff
@@ -63,24 +63,33 @@ const (
 	DropIndex      EventType = "drop index"
 	CreateView     EventType = "create view"
 	DropView       EventType = "drop view"
-	AlertTable     EventType = "alter table"
+	AlterTable     EventType = "alter table"
+
+	CreateSchema EventType = "create schema" // alias of CreateDatabase
+	DropSchema   EventType = "drop schema"   // alias of DropDatabase
```
Comment (lance6716) [inline pkg/binlog-filter/filter.go:70]: there's no AlterSchema/AlterDatabase 🤔

File: pkg/binlog-filter/filter.go [hunk 5]
```diff
@@ -115,14 +124,42 @@ func (b *BinlogEventRule) Valid() error {
...
 	}
-
 	return nil
 }
 
+// ErrInvalidEventType is returned when event type is not valid.
+var ErrInvalidEventType = errors.New("event type not found")
+
+// toEvent converts event type string to EventType and check if it is valid.
+func (b *BinlogEventRule) toEvent(es string) (EventType, error) {
```
Comment (lance6716) [inline pkg/binlog-filter/filter.go]: why this function is a method of `BinlogEventRule`? maybe you want iterate through b.Events inside this function?

Comment (asddongmen) [reply pkg/binlog-filter/filter.go]: good catch, it should not be a method of `BinlogEventRule`.

File: pkg/binlog-filter/filter.go (hunk 3 above)
Comment (asddongmen) [reply pkg/binlog-filter/filter.go]: Sure.

Comment (asddongmen) [reply pkg/binlog-filter/filter.go]: sorry, I'am wrong. We still need `NoneEvent ` to cover unsupported `ast.Stmt`.

File: pkg/binlog-filter/filter.go (hunk 4 above)
Comment (asddongmen) [reply pkg/binlog-filter/filter.go:70]: I have added. PTAL.

File: pkg/binlog-filter/filter.go [hunk 6]
```diff
@@ -63,24 +63,35 @@ const (
 	DropIndex      EventType = "drop index"
 	CreateView     EventType = "create view"
 	DropView       EventType = "drop view"
-	AlertTable     EventType = "alter table"
+	AlterTable     EventType = "alter table"
+
+	CreateSchema  EventType = "create schema" // alias of CreateDatabase
+	DropSchema    EventType = "drop schema"   // alias of DropDatabase
+	AddIndex      EventType = "add index"     // alias of CreateIndex
+	AlterDatabase EventType = "alter database"
```
Comment (lance6716) [inline pkg/binlog-filter/filter.go]: it should be moved to near DropDatabase

File: pkg/binlog-filter/filter.go (hunk 3 above)
Comment (lance6716) [reply pkg/binlog-filter/filter.go]: what's the defference of NullEvent, NoneEvent, InvalidEvent ? please add comments in code

File: pkg/binlog-filter/filter_test.go [hunk 7]
```diff
@@ -245,3 +246,33 @@ func (t *testFilterSuite) TestGlobalFilter(c *C) {
...
 }
+
+func (t *testFilterSuite) TestToEventType(c *C) {
+	cases := []struct {
+		eventStr string
+		event    EventType
+		err      error
+	}{
+		{"", NullEvent, nil},
+		{"insert", InsertEvent, nil},
```
Comment (lance6716) [inline pkg/binlog-filter/filter_test.go:258]: also test some uppercase input

File: pkg/binlog-filter/util.go [hunk 8]
```diff
@@ -38,10 +42,52 @@ func AstToDDLEvent(node ast.StmtNode) EventType {
 	case *ast.DropIndexStmt:
 		return DropIndex
 	case *ast.AlterTableStmt:
-		return AlertTable
+		return AlterTable
 	case *ast.CreateViewStmt:
 		return CreateView
+	case *ast.AlterDatabaseStmt:
```
Comment (lance6716) [inline pkg/binlog-filter/util.go:49]: move to near XXXDatabaseStmt

File: pkg/binlog-filter/filter_test.go (hunk 7 above)
Comment (asddongmen) [reply pkg/binlog-filter/filter_test.go:258]: done

File: pkg/binlog-filter/filter_test.go [hunk 9]
```diff
@@ -245,3 +247,39 @@ func (t *testFilterSuite) TestGlobalFilter(c *C) {
...
+		{"add index", CreateIndex, nil},
+		{"create index", CreateIndex, nil},
+		{"xxx", NullEvent, errors.NotValidf("event type %s", "xxx")},
+		{"I don't know", NullEvent, errors.NotValidf("event type %s", "I don't know")},
+	}
+
+	for _, cs := range cases {
+		event, err := toEventType(cs.eventStr)
+		c.Assert(cs.event, Equals, event)
+		if err != nil {
```
Comment (lance6716) [inline pkg/binlog-filter/filter_test.go:281]: when err == nil, should also check cs.err is also nil

Comment (asddongmen) [reply pkg/binlog-filter/filter_test.go:281]: done

**Review by lance6716 — APPROVED** [review]

**Pull request merged**